### PR TITLE
Chris Martin – Solution (Performance)

### PR DESCRIPTION
**Pull request by schrismartin**
This is my solution for the problem that passes the measure tests, but doesn't attempt to add functionality to the protocols via extensions.

### CODE REVIEW

File: Challenge Source/LinkedListCodingChallenge/Node.swift [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+//  Copyright © 2018 nashcocoaheads. All rights reserved.
+//
+
+import Foundation
+
+public final class Node<Element>: NodeType {
+    
+    public var element: Element
+    public var next: Node<Element>?
+    public var previous: Node<Element>?
```
Comment (schrismartin) [inline Challenge Source/LinkedListCodingChallenge/Node.swift:15]: Should be `weak`!

**Comment by patgoley** [comment]
Looks great! Your implementation has the advantage of being doubly linked, so it wouldn't be hard to add a reverse iterator. I believe your extension with `Iterator` and `makeIterator` could be moved to an extension on the protocol itself. I don't think that the `underestimatedCount` part is needed, though not sure.

**Comment by schrismartin** [comment]
I added the `underestimatedCount` because I didn't know if `for ... in` loops inherently took the `LinkedList`'s `underestimatedCount` into consideration here. This just adds an extra layer of safety.

**Comment by patgoley** [comment]
Interesting, I didn't realize that was defined on `Sequence`. [It seems like](https://stackoverflow.com/questions/39112934/underestimatedcount-in-swifts-dictionary) it's used for allocating memory when copying a sequence, so definitely a good optimization to have. Looping stops when the iterator returns nil. Either way, should be fine to leave and still move to the protocol extension if desired.

**Comment by schrismartin** [comment]
Because this implementation is doubly-linked, the nodes perform a circle. So if that check isn't there, it could potentially infinitely cycle.

That is to say, this implementation would never return nil.

**Pull request merged**